### PR TITLE
Bug/accessibility 5

### PR DESCRIPTION
**Pull request by udeeshagautam**
Fix for

#6946
Schema compare options dialog tabs are not accessible through tabbing. This is because options dialogs don't gain focus when opened) Fixed in core. Tested with core and Ext dialogs.

#6796 (second part)
Data cells (text data mostly) with no data in table should announce no data available. Ensuring that this doesn't affect checkbox cells.
 
#6726
Checkboxes should announce 'checkbox' and their status

https://github.com/microsoft/azuredatastudio/issues/6852
Focus wasn't visible when moving in grid in HC since the whole row is selected in same color. Added stronger border for active cell.

### CODE REVIEW

File: src/sql/base/browser/ui/table/formatters.ts [hunk 1]
```diff
@@ -77,6 +77,19 @@ export function slickGridDataItemColumnValueExtractor(value: any, columnDef: any
 	};
 }
 
+/**
+ * Alterate function to provide slick grid cell with area label and plain text
+ * In this case, for no display value arialable will be set to specific string "no data available" for accessibily support for screen readers
```
Comment (kisantia) [inline src/sql/base/browser/ui/table/formatters.ts]: nit: arialabel

Comment (udeeshagautam) [reply src/sql/base/browser/ui/table/formatters.ts]: ouch sorry about that

File: src/sql/platform/dialog/browser/dialogModal.ts [hunk 2]
```diff
@@ -110,12 +110,12 @@ export class DialogModal extends Modal {
 		dialogButton.hidden ? buttonElement.element.parentElement.classList.add('dialogModal-hidden') : buttonElement.element.parentElement.classList.remove('dialogModal-hidden');
 	}
 
-	protected renderBody(container: HTMLElement): void {
+	protected renderBody(container: HTMLElement): HTMLElement {
```
Comment (anthonydresser) [inline src/sql/platform/dialog/browser/dialogModal.ts]: We shouldn't need to return this, the modal class has access to the `container` being passed in here.

**Comment by kisantia** [comment]
Does this also fix #6728 since the focus is on the dialog now when it opens?

File: src/sql/workbench/browser/modal/modal.ts [hunk 3]
```diff
@@ -234,7 +235,7 @@ export abstract class Modal extends Disposable implements IThemable {
 		const modalBodyClass = (this._modalOptions.isAngular === false ? 'modal-body' : 'modal-body-and-footer');
 
 		this._modalBodySection = DOM.append(modalContent, DOM.$(`.${modalBodyClass}`));
-		this.renderBody(this._modalBodySection);
+		this._dialogPaneBody = <HTMLElement>this.renderBody(this._modalBodySection);
```
Comment (anthonydresser) [inline src/sql/workbench/browser/modal/modal.ts]: why don't we just use `this._modalBodySection` anywhere we are want to search the body?

File: src/sql/workbench/browser/modal/modal.ts [hunk 4]
```diff
@@ -252,7 +253,7 @@ export abstract class Modal extends Disposable implements IThemable {
 	 * Called for extended classes to render the body
 	 * @param container The parent container to attach the rendered body to
 	 */
-	protected abstract renderBody(container: HTMLElement): void;
+	protected abstract renderBody(container: HTMLElement): HTMLElement | void;
```
Comment (anthonydresser) [inline src/sql/workbench/browser/modal/modal.ts]: same comment

File: src/sql/workbench/browser/modal/modal.ts [hunk 5]
```diff
@@ -319,13 +320,21 @@ export abstract class Modal extends Disposable implements IThemable {
 	 * Set focusable elements in the modal dialog
 	 */
 	public setFocusableElements() {
-		this._focusableElements = this._bodyContainer.querySelectorAll('a[href], area[href], input:not([disabled]), select:not([disabled]), textarea:not([disabled]), button:not([disabled]), [tabindex="0"]');
+		// try to find focusable element in dialog pane rather than overall container
+		this._focusableElements = this._dialogPaneBody ?
```
Comment (anthonydresser) [inline src/sql/workbench/browser/modal/modal.ts]: does it make more sense to just call focus on the modal itself? That way implementers have the chance to define this behavior?

File: src/sql/base/browser/ui/table/formatters.ts [hunk 6]
```diff
@@ -77,6 +77,19 @@ export function slickGridDataItemColumnValueExtractor(value: any, columnDef: any
 	};
 }
 
+/**
+ * Alterate function to provide slick grid cell with area label and plain text
```
Comment (Benjin) [inline src/sql/base/browser/ui/table/formatters.ts]: typo "alternate"

File: src/sql/base/browser/ui/table/formatters.ts [hunk 7]
```diff
@@ -77,6 +77,19 @@ export function slickGridDataItemColumnValueExtractor(value: any, columnDef: any
...
+ * In this case, for no display value arialable will be set to specific string "no data available" for accessibily support for screen readers
+ * Set 'no data' lable only if cell is present and has no value (so that checkbox and other custom plugings do not get 'no data' label)
+ */
+export function slickGridDataItemColumnValueWithNoData(value: any, columnDef: any): { text: string; ariaLabel: string; } {
+	let displayValue = value[columnDef.field];
+	return {
+		text: displayValue,
+		ariaLabel: displayValue ? escape(displayValue) : ((displayValue !== undefined) ? 'no data available' : displayValue)
```
Comment (Benjin) [inline src/sql/base/browser/ui/table/formatters.ts]: localized?

File: src/sql/base/browser/ui/table/plugins/checkboxSelectColumn.plugin.ts [hunk 8]
```diff
@@ -243,32 +243,37 @@ export class CheckboxSelectColumn<T extends Slick.SlickData> implements Slick.Pl
...
-			return strings.format(checkboxTemplate, '', this._options.title);
+			return strings.format(checkboxTemplate, '', this.getariaLabel(false));
 		}
 	}
 
 	private isCustomActionRequested(): boolean {
 		return (this._options.actionOnCheck === ActionOnCheck.customAction);
 	}
+
+	private getariaLabel(checked: boolean): string {
```
Comment (Benjin) [inline src/sql/base/browser/ui/table/plugins/checkboxSelectColumn.plugin.ts]: nit: capitalize 'A' in "aria"

File: src/sql/base/browser/ui/table/plugins/checkboxSelectColumn.plugin.ts [hunk 9]
```diff
@@ -243,32 +243,37 @@ export class CheckboxSelectColumn<T extends Slick.SlickData> implements Slick.Pl
...
+		return checked ? strings.format('"{0} {1}"', this._options.title, 'checkbox checked') :
```
Comment (Benjin) [inline src/sql/base/browser/ui/table/plugins/checkboxSelectColumn.plugin.ts]: localize "checkbox checked"/"checkbox unchecked" strings

File: src/sql/base/browser/ui/table/formatters.ts [hunk 10]
```diff
@@ -77,6 +77,19 @@ export function slickGridDataItemColumnValueExtractor(value: any, columnDef: any
 	};
 }
 
+/**
+ * Alterate function to provide slick grid cell with area label and plain text
+ * In this case, for no display value arialable will be set to specific string "no data available" for accessibily support for screen readers
+ * Set 'no data' lable only if cell is present and has no value (so that checkbox and other custom plugings do not get 'no data' label)
```
Comment (Benjin) [inline src/sql/base/browser/ui/table/formatters.ts]: typo "plugins"

File: src/sql/workbench/browser/modal/modal.ts (hunk 3 above)
Comment (udeeshagautam) [reply src/sql/workbench/browser/modal/modal.ts]: yeah I tried that first - but the element return its not dialog pane

**Comment by udeeshagautam** [comment]
@kisantia yes seems like it does :)

File: src/sql/base/browser/ui/table/plugins/checkboxSelectColumn.plugin.ts [hunk 11]
```diff
@@ -243,32 +243,37 @@ export class CheckboxSelectColumn<T extends Slick.SlickData> implements Slick.Pl
...
+			return strings.format(checkboxTemplate, '', this.getAriaLabel(false));
 		}
 	}
 
 	private isCustomActionRequested(): boolean {
 		return (this._options.actionOnCheck === ActionOnCheck.customAction);
 	}
+
+	private getAriaLabel(checked: boolean): string {
+		return checked ? strings.format('"{0} {1}"', this._options.title, nls.localize("tableCheckboxCell.Checked", "checkbox checked")) :
```
Comment (Benjin) [inline src/sql/base/browser/ui/table/plugins/checkboxSelectColumn.plugin.ts]: why not use a template string instead?

`strings.format('"{0} {1}"', this._options.title, nls.localize("tableCheckboxCell.Checked", "checkbox checked")) :`
becomes
`'"${this._options.title} ${nls.localize("tableCheckboxCell.Checked", "checkbox checked")}"'`

**Pull request merged**